vql/linux/cronsnoop/cron-lib.go

```diff
@@ -0,0 +1,395 @@
+}
+
+func (snooper *CronSnooper) is_system_cron_event(event fsnotify.Event) bool {
+	for _, prefix := range snooper.cron_system_paths {
+		if strings.HasPrefix(event.Name, prefix) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func exists(haystack []string, needle string) bool {
+	for _, cmd := range haystack {
+		if cmd == needle {
+			return true
+		}
+	}
+
+	return false
+}
+
+// returns a slice, containing only those entries which are both in a and b,
+// entries existing in a, but not in b are removed from a
+func filter_deleted_cmds(a, b []string) ([]string, []string) {
+	// filtered re-uses the storage of cmds
+	existing := a[:0]
+	deleted := []string{}
+	for _, cmd := range a {
+		if exists(b, cmd) {
+			existing = append(existing, cmd)
+		} else {
+			deleted = append(deleted, cmd)
+		}
+
+	}
+
+	// ensure everything which got filtered in a is garbage collected
+	for i := len(existing); i < len(a); i++ {
+		a[i] = ""
+	}
+
+	return existing, deleted
+}
+
+func special_cron_string(line_fields []string) bool {
+	return line_fields[0][0] == '@'
+}
+
+func should_skip_line(line_fields []string) bool {
+	// ignore comments and empty lines
+	if line_fields[0][0] == '#' || len(line_fields) == 0 {
+		return true
+	}
+
+	// if line starts with @ special string instead of ordinary cron time/date
+	// fields don't skip it.
+	if special_cron_string(line_fields) {
+		return false
+	} else if len(line_fields) <= 5 {
+		return true
+	}
+
+	return false
+}
+
+func (snooper *CronSnooper) parse_user_cron_file(path string) error {
+
+	var existing_user_cmds, current_commands []string
+	var cmd string
+	var ok bool
+
+	user := filepath.Base(path)
+
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+
+	// initialize user entries on first user modification detectioon
+	if existing_user_cmds, ok = snooper.user_cron_registry[user]; !ok {
+		snooper.user_cron_registry[user] = existing_user_cmds
+	}
+
+	for scanner.Scan() {
+		line := strings.TrimLeft(scanner.Text(), " \t")
+		split_line := strings.Split(line, " ")
+
+		if should_skip_line(split_line) {
+			continue
+		}
+
+		// Handle @-based entries in cron file
+		if special_cron_string(split_line) {
+			cmd = strings.Join(split_line[1:], " ")
+		} else {
+			cmd = strings.Join(split_line[5:], " ")
+		}
+
+		current_commands = append(current_commands, cmd)
+		// only take note of unique commands
+		if !exists(existing_user_cmds, cmd) {
+			existing_user_cmds = append(existing_user_cmds, cmd)
+			snooper.emit_event(cmd, user, path, Added)
+		}
+	}
+
+	// Filter out deleted entries
+	existing, deleted := filter_deleted_cmds(existing_user_cmds, current_commands)
+
+	// emit events
+	for _, deleted_cmd := range deleted {
+		snooper.emit_event(deleted_cmd, user, path, Deleted)
+	}
+
+	// user got all of its entries deleted so just delete the user altogether
+	if len(existing) == 0 {
+		//TODO: Emit an event that all of user's entries are deleted
```

Comment:
Is this a TODO you want to do before merge?

Reply:
So looking at the todo and the surrounding code all the of user's commands would have already been emitted as being deleted due to :

```
   // emit events                                                              
    for _, deleted_cmd := range deleted {                                       
        snooper.emit_event(deleted_cmd, user, path, Deleted)                    
    }  
```

So even if I don't do anything for the todo we are not missing any information w.r.t what entries got deleted, however we won't get anything explicitly saying that all commands got deleted. Imo this is not a big deal. What do you say ?